ibis/expr/operations/arrays.py

```diff
@@ -7,23 +7,32 @@
 import ibis.expr.datashape as ds
 import ibis.expr.datatypes as dt
 import ibis.expr.rules as rlz
-from ibis.common.annotations import attribute
+from ibis.common.annotations import ValidationError, attribute
 from ibis.common.typing import VarTuple  # noqa: TCH001
 from ibis.expr.operations.core import Unary, Value
 
 
 @public
 class Array(Value):
     exprs: VarTuple[Value]
+    dtype: Optional[dt.Array] = None
+
+    def __init__(self, exprs, dtype: dt.Array | None = None):
+        if len(exprs) == 0:
+            if dtype is None:
+                raise ValidationError("If values is empty, dtype must be provided")
+            if not isinstance(dtype, dt.Array):
+                raise ValidationError(f"dtype must be an array, got {dtype}")
```

Comment:
You don't need this at all, the declaration of `dtype: Optional[dt.Array]` creates a validation rule.

Reply:
You are right it has a validation rule, but the generic error message isn't that clear.

Also just to be clear (don't think it matters here) the validation from the signature happens during the `super().__init__()` call, so the validation will happen, but in the future, it hasn't happened yet.